packages/SwingSet/src/vats/network/network.js

```diff
@@ -0,0 +1,436 @@
+ * @property {(port: Port, local: Endpoint, remote: Endpoint, l: ListenHandler) => Promise<ConnectionHandler>} [onAccept] A new connection is incoming
+ * @property {(port: Port, rej: any, l: ListenHandler) => Promise<void>} [onError] There was an error while listening
+ * @property {(port: Port, l: ListenHandler) => Promise<void>} [onRemove] The listener has been removed
+ */
+
+/**
+ * @typedef {Object} Connection
+ * @property {(packetBytes: Data) => Promise<Bytes>} send Send a packet on the connection
+ * @property {() => void} close Close both ends of the connection
+ */
+
+/**
+ * @typedef {Object} ConnectionHandler A handler for a given Connection
+ * @property {(connection: Connection, c: ConnectionHandler) => void} [onOpen] The connection has been opened
+ * @property {(connection: Connection, packetBytes: Bytes, c: ConnectionHandler) => Promise<Data>} [onReceive] The connection received a packet
+ * @property {(connection: Connection, reason?: CloseReason, c: ConnectionHandler) => Promise<void>} [onClose] The connection has been closed
+ *
+ * @typedef {any?} CloseReason The reason a connection was closed
+ */
+
+/**
+ * @typedef {Object} Packet
+ * @property {Endpoint} src Source of the packet
+ * @property {Endpoint} dst Destination of the packet
+ * @property {Bytes} bytes Bytes in the packet
+ * @property {import('@agoric/produce-promise').PromiseRecord<Bytes>} deferredAck The deferred to resolve the result
+ *
+ * @typedef {number} TTL A time-to-live for a packet
+ */
+
+/**
+ * @typedef {Object} InterfaceHandler A handler for things the interface implementation will invoke
+ * @property {(interface: InterfaceImpl, p: InterfaceHandler) => Promise<void>} onCreate This interface is created
+ * @property {(port: Port, localAddr: Endpoint, listenHandler: ListenHandler, p: InterfaceHandler) => Promise<void>} onListen A port was listening
+ * @property {(port: Port, localAddr: Endpoint, listenHandler: ListenHandler, p: InterfaceHandler) => Promise<void>} onListenRemove A port listener has been reset
+ * @property {(port: Port, localAddr: Endpoint, remote: Endpoint, p: InterfaceHandler) => Promise<ConnectionHandler>} onConnect A port initiates an outbound connection
+ * @property {(port: Port, localAddr: Endpoint, p: InterfaceHandler) => Promise<void>} onRevoke The port is being completely destroyed
+ *
+ * @typedef {Object} InterfaceImpl Things the interface can do for us
+ * @property {(port: Port, remote: Endpoint, connectionHandler: ConnectionHandler) => Promise<Connection>} connect Establish a connection from this interface to an endpoint
+ */
+
+/*
+ * Convert some data to bytes.
+ *
+ * @param {Data} data
+ * @returns {Bytes}
+ */
+export function toBytes(data) {
+  // We really need marshallable TypedArrays.
+  if (typeof data === 'string') {
+    // eslint-disable-next-line no-bitwise
+    data = data.split('').map(c => c.charCodeAt(0));
```

Comment:
eww.. is that to avoid relying on Node's `Buffer` type to do the encoding? I suppose I'm ok with that as a goal, but I expect we'll fix `marshal` to handle bytes before we get the rest of this working in a non-Node environment.

(oh, but, I'm not even sure we provide `Buffer` inside SES compartments, so maybe this is the only way to go)

Reply:
Yup.  And the hack is documented in `networking.md`.